pennylane/gradients/param_shift_hessian.py

```diff
@@ -0,0 +1,298 @@
+                hessian_coeffs.append([])
+                shapes.append(((i, j), 0))
+                continue
+
+            recipe = diag_recipe if i == j else off_diag_recipe
+            coeffs, _, shifts = _process_hessian_recipe(recipe)
+
+            # optimization: only compute the unshifted tape once
+            if all(np.array(shifts[0]) == 0):
+                if not unshifted_coeffs and f0 is None:
+                    hessian_tapes.insert(0, tape)
+
+                unshifted_coeffs[(i, j)] = coeffs[0]
+                coeffs, shifts = coeffs[1:], shifts[1:]
+
+            # generate the Hessian tapes
+            hessian_coeffs.append(coeffs)
+            h_tapes = generate_multishifted_tapes(tape, (i, j), shifts)
+
+            hessian_tapes.extend(h_tapes)
+            shapes.append(((i, j), len(h_tapes)))
+
+    def processing_fn(results):
+        # The first results dimension is the number of terms/tapes in the parameter-shift
+        # rule, the remaining ones are the QNode output dimensions.
+        out_dim = qml.math.shape(qml.math.stack(results))[1:]
+        # The desired shape of the Hessian is:
+        #       (QNode output dimensions, # trainable gate args, # trainable gate args),
+        # but first we accumulate all elements into a list, since no array assingment is possible.
+        hessian = []
+        # Keep track of tape results already consumed.
+        start = 1 if unshifted_coeffs and f0 is None else 0
+        # Results of the unshifted tape.
+        r0 = results[0] if start == 1 else f0
+
+        for k, ((i, j), s) in enumerate(shapes):
+            res = results[start : start + s]
+            start = start + s
+
+            # Compute the elements of the Hessian as the linear combination of
+            # results and coefficients, barring optimization cases.
+            if j < i:
+                g = hessian[j * h_dim + i]
+            elif s == 0:
+                g = qml.math.zeros(out_dim)
+            else:
+                res = qml.math.stack(res)
+                g = qml.math.tensordot(
+                    res, qml.math.convert_like(hessian_coeffs[k], res), [[0], [0]]
+                )
+                if (i, j) in unshifted_coeffs:
+                    g += unshifted_coeffs[(i, j)] * r0
+
+            hessian.append(g)
+
+        # Reshape the Hessian to have the dimensions of the QNode output on the outside, that is:
+        #         (h_dim, h_dim, out_dim) -> (out_dim, h_dim, h_dim)
+        hessian = qml.math.reshape(qml.math.stack(hessian), (h_dim, h_dim) + out_dim)
+        reordered_axes = list(range(2, len(out_dim) + 2)) + [0, 1]
+        hessian = qml.math.transpose(hessian, axes=reordered_axes)
+
+        return qml.math.squeeze(hessian)
+
+    return hessian_tapes, processing_fn
+
+
+@hessian_transform
+def param_shift_hessian(tape, f0=None):
+    r"""Transform a QNode to compute the parameter-shift Hessian with respect to its trainable
+    parameters.
+
```

Comment:
@dime10 just to make sure that we are explicit for now, it might be good to add a `.. note::` or a `.. warning::` here just to note that this function only supports gates with 2-term rules, and will raise an error otherwise.